include/spi.h

```diff
@@ -27,173 +37,182 @@
+ *    word_size           [ 5 : 10 ]  - Size of a data frame in bits.
+ *    lines               [ 11 : 12 ] - MISO lines: Single/Dual/Quad.
+ *    RESERVED            [ 13 : 15 ] - Unused yet
+ *
+ * slave is the slave number from 0 to host constoller slave limit.
+ *
+ * cs is a valid pointer on a struct spi_cs_control is CS line is
+ *    emulated through a gpio line, or NULL otherwise.
  */
-typedef int (*spi_api_configure)(struct device *dev,
-				 struct spi_config *config);
+struct spi_config {
+	u32_t		frequency;
+	u16_t		operation;
+	u16_t		slave;
+	struct spi_cs_control *cs;
+};
+
 /**
- * @typedef spi_api_slave_select
- * @brief Callback API upon selecting a slave
- * See spi_slave_select() for argument description
+ * @brief SPI buffer structure
+ *
+ * buf is a valid pointer on a data buffer, or NULL otherwise.
+ * len is the length of the buffer or, if buf is NULL, will be the
+ *     length which as to be sent as dummy bytes (as TX buffer) or
+ *     the length of bytes that should be skipped (as RX buffer).
  */
-typedef int (*spi_api_slave_select)(struct device *dev, u32_t slave);
+struct spi_buf {
+	void *buf;
+	u32_t len;
+};
+
 /**
  * @typedef spi_api_io
  * @brief Callback API for I/O
- * See spi_read() and spi_write() for argument descriptions
+ * See spi_transceive() for argument descriptions
  */
-typedef int (*spi_api_io)(struct device *dev,
-			  const void *tx_buf, u32_t tx_buf_len,
-			  void *rx_buf, u32_t rx_buf_len);
+typedef int (*spi_api_io)(struct device	*dev,
+			  struct spi_config *config,
+			  const struct spi_buf **tx_bufs,
+			  struct spi_buf **rx_bufs);
 
+/**
+ * @brief SPI driver API
+ * This is the mandatory API any SPI driver needs to expose.
+ */
 struct spi_driver_api {
-	spi_api_configure configure;
-	spi_api_slave_select slave_select;
 	spi_api_io transceive;
 };
 
 /**
- * @brief Configure a host controller for operating against slaves.
- * @param dev Pointer to the device structure for the driver instance.
- * @param config Pointer to the configuration provided by the application.
+ * @brief Read/write the specified amount of data from the SPI driver.
  *
- * @retval 0 If successful.
- * @retval Negative errno code if failure.
- */
-static inline int spi_configure(struct device *dev,
-				struct spi_config *config)
-{
-	const struct spi_driver_api *api = dev->driver_api;
-
-	return api->configure(dev, config);
-}
-
-/**
- * @brief Select a slave to deal with.
+ * Note: This function is synchronous.
  *
- * This routine is meaningful only if the controller supports per-slave
- * addressing: One SS line per-slave. If not, this routine has no effect
- * and daisy-chaining should be considered to deal with multiple slaves
- * on the same line.
+ * @param dev is a valid pointer to an actual SPI device
+ * @param config Pointer to a valid spi_config structure instance.
+ * @param tx_bufs Memory buffer(s) where data to be sent originates from.
```

Comment:
Looks like a good improvement.

Are tx_bufs and rx_bufs null-terminated? If so, could this be documented?